### ISSUE TITLE
Better source of localization key

### ISSUE THREAD

**Issue by svmnotn** [TeamPorcupine/ProjectPorcupine#216]
Currently the localization system requires extra tags in the furniture, it would be simple to use the name, and description tags for this purpose so as to not have any duplication as the en_US file is currently just a duplicate of some of the Furniture.xml fields. However the Name tag at least is also used for the textures as far as I know.

Thus if we were to change there would need to be a massive renaming of all the code that is currently hardcoded to somthing like `== "Door"` to use the new name (or preferably not do that at all, maybe expose that to lua and let it handle that?).

So here is the list of places that would need to be renamed:

- [ ] Furniture.xml
- [ ] everything within Images/Furniture

I know there are other places but I would like some help finding them all, that is if we all agree that this would be a better way of handling the localisations.

To recap, I'm proposing we do the following: 

1. Delete `LocalizationCode` and `UnlocalizedDescription` from Furniture.xml
2. Replace the text inside `Name` and `Description` with what was in the removed tags
3. Update all other references to the old names and perhaps improve whatever was using them to not include hardcoded names.

---
As an aside this would be a great example of an RFC in my opinion.

**Comment by BeefEX** [comment]
I am working on animation system that loads images based on their objectName. In order for my system to work filenames of images need to be same as objectName of that furniture.

**Comment by vogonistic** [comment]
I was thinking that we keep `Name` and `Description` as is and use with `Furniture.objectType` as the base key for identification and localization.

**Comment by BeefEX** [comment]
@vogonistic thats very good solution

**Comment by svmnotn** [comment]
@vogonistic I didn't even see objType, but the code for the localization would not fall back to the name as it does not know anything about the furniture.

So in regards to the current system objName is only used for the mouse over text, which is not localized (the build buttons only use the localization key which I think is what caused #209). However if we made it so that it is used as the key of the localization then we would get something like this:

Furniture.xml \/
```xml
<Furniture objectType="furn_SteelWall">
        <Name>furn_steel_wall</Name>
        <Description>furn_steel_wall_desc</Description>
</Furniture>
```
en_US.lang \/
```
furn_steel_wall=Steel Wall
furn_steel_wall_desc=A wall made out of steel.
```
resulting in: `Build Steel Wall`

And something like this \/
```xml
<Furniture objectType="furn_SteelWall">
        <Name>Steel Wall</Name>
        <Description>A wall made out of steel.</Description>
</Furniture>
```
Also resulting in `Build Steel Wall`

However if we use the objType then this \/
```xml
<Furniture objectType="furn_SteelWall">
        <Name>Steel Wall</Name>
        <Description>A wall made out of steel.</Description>
</Furniture>
```
results in `Build furn_SteelWall`

**Comment by svmnotn** [comment]
that said is `MouseOverFurnitureTypeText` even used anymore?

**Comment by vogonistic** [comment]
Unless the signature for getting a localization is something like this:
```xml
<Furniture objectType="furn_SteelWall">
        <Name>Steel Wall</Name>
        <Description>A wall made out of steel.</Description>
</Furniture>
```

```cs
public GetLocalizedString(string key, string fallback) {}
Furniture f = ThatSteelWallXml();

// Results in "Steel Wall" if there is no localized value for "furn_SteelWall".
string name = GetLocalizedString(f.objectType, f.Name);
// Results in "A wall made out of steel." if there is no localized value for "furn_SteelWall.description".
string desc = GetLocalizedString(f.objectType+".description", f.Description);
```

This way the Furniture.xml works without any external file to give default values. 

---

@svmnotn Is your suggestion that we apply localization when loading the Furniture.xml? If so, one way to make sure that `en_US.lang` contains fallback values for all keys would be to prevent the game from starting if it's running in the editor and missing. I just want to make it obvious what we expect out of the modders and creators so that we have simple sane default values and allow them to easily figure out what to override for translation.

In both of our solutions, it would be easy to write a script that extracts all keys that needs to be translated.

**Comment by svmnotn** [comment]
The change of signature requires more parameters to be passed in, not to sound lazy but in my experience the more parameters the function has the more issues there are going to be.

@vogonistic no. My suggestion is that we use the name as the localization key, if it is not found in the .lang file then it will be used 'as is' in the xml.

**Comment by vogonistic** [comment]
The only reason I'm not onboard is because I prefer not to require devs/modders to maintain two files in parallel.

**Comment by svmnotn** [comment]
Ok.. now you've lost me how would it require that? (in difference to your proposal?)

**Comment by vogonistic** [comment]
My proposal means that `Furniture.xml` contains everything to display the en_US default values without having a `en_US.lang` file. In your proposal, `furn_steel_wall` is displayed if the person who added the steel wall didn't also update the `en_US.lang` file. Am I misunderstanding your proposal?

**Comment by svmnotn** [comment]
@vogonistic ah, I got it now. That is true, I'm just not a fan of having to add extra information to the localization code. However that would work just fine, how would you get the description? Just add _desc to the objType?

**Comment by svmnotn** [comment]
actually to make life a tad easier, how about we change Furniture's implementation of the ISelectable interface so that it returns the "human" readable name instead of the localization key so that #209 won't happen?

So something like:
```c#
public string GetName()
{
        return LocalizationTable.GetLocalization(this.objectType, this.Name);
}
```

**Comment by vogonistic** [comment]
I like the GetName() implementation, but pluralization might be required. 

The remaining problem is localizing the menus and UI. Not sure how that works.

**Comment by svmnotn** [comment]
@vogonistic pluralization depends on the language too much, maybe add an extra `GetPluralName` method?

Well the menus and the UI just use their text as the localization key (from what I saw before running the game)

**Comment by svmnotn** [comment]
@vogonistic I'll be closing this and taking a look at the source, if something like what we've come up here is not done (I'll try to check PRs but there are soo many) then I'll make a PR with the relevant changes.
